README.md
# sider_bar

sider bar for listview like ios contact list

## Usage

### Import the package
First, add sider_bar as a dependency in your pubspec.yaml

Then, import it:
```dart
import 'package:sider_bar/sider_bar.dart';
```
#### wrap ListView in a Stack,and SideBar over ListView with AlignmentDirectional.centerEnd
```dart
@override
  Widget build(BuildContext context) {
    return new Scaffold(
        appBar: new AppBar(
          // Here we take the value from the MyHomePage object that was created by
          // the App.build method, and use it to set our appbar title.
          title: new Text(widget.title),
        ),
        body: Stack(
          alignment: AlignmentDirectional.centerEnd,
          children: <Widget>[
            ListView.builder(
              controller: _controller,
              itemCount: mDataList.length,
              itemBuilder: buildItem,
            ),
            SideBar(
                list: mDataList,
                textColor: Colors.blue,
                color: Colors.blue.withOpacity(0.2),
                valueChanged: (value) {
                  _controller
                      .jumpTo(mDataList.indexOf(value) * 44.0); //card 差不多44的高度
                })
          ],
        ),
    );
  }
```

### capture

![示例](capture.gif)


CHANGELOG.md
## [0.1.0] - 2018.09,18

* add: sidebar引入.


test/sider_bar_test.dart
//import 'package:test/test.dart';
//
//import 'package:sider_bar/sider_bar.dart';
//
//void main() {
//  test('adds one to input values', () {
//  });
//}


lib/sider_bar.dart
import 'package:flutter/material.dart';
import 'package:flutter/cupertino.dart';

///列表右侧的滑块
///textColor 文字颜色
///color 按住之后的颜色
///valueChanged 回调
///
class SideBar extends StatefulWidget {
  final List<String> list;
  final Color textColor;
  final Color color;
  final ValueChanged<String> valueChanged;

  SideBar(
      {Key key,
      @required this.list,
      @required this.textColor,
      @required this.color,
      @required this.valueChanged})
      : assert(list != null || list.length > 0),
        assert(valueChanged != null),
        super(key: key);

  @override
  _SideBarState createState() => _SideBarState();
}

class _SideBarState extends State<SideBar> {
  List<Widget> _widgetList;
  bool _isTaped = false;
  String _selectStr = "";

  @override
  void initState() {
    // TODO: implement initState
    super.initState();
    _widgetList = List<Widget>.generate(widget.list.length, (index) {
      return Padding(
        padding: const EdgeInsets.only(bottom: 2.0),
        child: Text(
          widget.list[index],
          style: TextStyle(color: widget.textColor, fontSize: 14.0),
        ),
      );
    });
  }

  _onDragStart(BuildContext context, DragStartDetails start) {
    RenderBox getBox = context.findRenderObject();
    var local = getBox.globalToLocal(start.globalPosition);
    setState(() {
      _selectStr = _getCurrentStr(local.dy);
      widget.valueChanged(_selectStr);
      _isTaped = true;
    });
  }

  _onDragUpdate(BuildContext context, DragUpdateDetails update) {
    RenderBox getBox = context.findRenderObject();
    var local = getBox.globalToLocal(update.globalPosition);
    setState(() {
      _selectStr = _getCurrentStr(local.dy);
      widget.valueChanged(_selectStr);
    });
  }

  @override
  Widget build(BuildContext context) {
    return GestureDetector(
      onVerticalDragStart: (DragStartDetails start) {
        _onDragStart(context, start);
      },
      onVerticalDragUpdate: (DragUpdateDetails update) =>
          _onDragUpdate(context, update),
      onVerticalDragEnd: (DragEndDetails end) {
        setState(() {
          _isTaped = false;
        });
      },
      child: Row(
        mainAxisSize: MainAxisSize.min,
        children: <Widget>[
          _isTaped
              ? Container(
                  alignment: Alignment.center,
                  margin: EdgeInsets.only(
                      right: MediaQuery.of(context).size.width / 2 - 70 + 15),
                  width: 70.0,
                  height: 70.0,
                  decoration: BoxDecoration(
                      color: widget.color,
                      shape: BoxShape.rectangle,
                      borderRadius: BorderRadius.all(Radius.circular(5.0))),
                  child: Text(
                    _selectStr,
                    style: Theme.of(context)
                        .textTheme
                        .display1
                        .copyWith(color: widget.textColor),
                  ),
                )
              : SizedBox(),
          Container(
            padding: EdgeInsets.symmetric(vertical: 8.0),
            decoration: BoxDecoration(
                color: _isTaped ? widget.color : Colors.transparent,
                shape: BoxShape.rectangle,
                borderRadius: BorderRadius.all(Radius.circular(15.0))),
            width: 26.0,
            child: Column(
              mainAxisSize: MainAxisSize.min,
              children: _widgetList,
            ),
          ),
        ],
      ),
    );
  }

  String _getCurrentStr(double dy) {
    int index = (dy / 19.0).round();
    if (index <= 0) {
      return widget.list.first;
    } else if (index >= widget.list.length) {
      return widget.list.last;
    } else {
      return widget.list[index];
    }
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:sider_bar/sider_bar.dart';

void main() => runApp(new MyApp());

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return new MaterialApp(
      title: 'Flutter Demo',
      theme: new ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or press Run > Flutter Hot Reload in IntelliJ). Notice that the
        // counter didn't reset back to zero; the application is not restarted.
        primarySwatch: Colors.blue,
      ),
      home: new MyHomePage(title: 'SideBar'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  // This widget is the home page of your application. It is stateful, meaning
  // that it has a State object (defined below) that contains fields that affect
  // how it looks.

  // This class is the configuration for the state. It holds the values (in this
  // case the title) provided by the parent (in this case the App widget) and
  // used by the build method of the State. Fields in a Widget subclass are
  // always marked "final".

  final String title;

  @override
  _MyHomePageState createState() => new _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  final String content = "ABCDEFGHIGKLMNOPQRSTUVWXYZ";
  ScrollController _controller = ScrollController();
  List<String> mDataList;

  @override
  void initState() {
    // TODO: implement initState
    super.initState();

    mDataList =
        List<String>.generate(content.length, (index) => content[index]);
  }

  @override
  Widget build(BuildContext context) {
    return new Scaffold(
        appBar: new AppBar(
          // Here we take the value from the MyHomePage object that was created by
          // the App.build method, and use it to set our appbar title.
          title: new Text(widget.title),
        ),
        body: Stack(
          alignment: AlignmentDirectional.centerEnd,
          children: <Widget>[
            ListView.builder(
              controller: _controller,
              itemCount: mDataList.length,
              itemBuilder: buildItem,
            ),
            SideBar(
                list: mDataList,
                textColor: Colors.blue,
                color: Colors.blue.withOpacity(0.2),
                valueChanged: (value) {
                  _controller
                      .jumpTo(mDataList.indexOf(value) * 44.0); //card 差不多44的高度
                })
          ],
        ),
    );
  }

  Widget buildItem(BuildContext context, int index) {
    return Card(
      child: ListTile(
        title: Text(mDataList[index]),
      ),
    );
  }
}
